packages/wrangler/src/pages/publish.tsx

```diff
@@ -327,47 +335,65 @@ export const Handler = async ({
 	}
 
 	if (builtFunctions) {
+		// with Pages Functions
+		// https://developers.cloudflare.com/pages/platform/functions/
 		formData.append("_worker.js", new File([builtFunctions], "_worker.js"));
 		logger.log(`✨ Uploading Functions`);
-		try {
-			_routes = readFileSync(routesOutputPath, "utf-8");
-			if (_routes) {
-				formData.append("_routes.json", new File([_routes], "_routes.json"));
+
+		if (_routesCustom) {
+			// user provided a custom _routes.json file
+			try {
+				validateRoutesFile(_routesCustom, join(directory, "_routes.json"));
+
+				formData.append(
+					"_routes.json",
+					new File([_routesCustom], "_routes.json")
+				);
+				logger.log(`✨ Uploading _routes.json`);
+				logger.warn(
+					`_routes.json is an experimental feature and is subject to change. Please use with care.`
+				);
+			} catch (err) {
+				if (err instanceof FatalError) {
+					throw err;
+				}
 			}
-		} catch {}
+		} else if (routesOutputPath) {
+			// no custom _routes.json file found, so fallback to the generated one
+			try {
+				_routesGenerated = readFileSync(routesOutputPath, "utf-8");
+
+				if (_routesGenerated) {
+					formData.append(
+						"_routes.json",
+						new File([_routesGenerated], "_routes.json")
+					);
+				}
+			} catch {}
+		}
 	} else if (_workerJS) {
 		// Advanced Mode
 		// https://developers.cloudflare.com/pages/platform/functions/#advanced-mode
 		formData.append("_worker.js", new File([_workerJS], "_worker.js"));
 		logger.log(`✨ Uploading _worker.js`);
 
-		try {
-			// In advanced mode, developers can specify a custom _routes.json
-			// file. In which case, we need to run it through optimization
-			// to potentially reduce the overall worker pipeline size
-			const routesPath = join(directory, "_routes.json");
-			const advancedModeRoutesString = readFileSync(routesPath, "utf-8");
-			const advancedModeRoutes = JSON.parse(advancedModeRoutesString);
-
-			if (!isRoutesJSONSpec(advancedModeRoutes)) {
-				throw new FatalError(
-					"Invalid _routes.json file found at:" + routesPath,
-					1
-				);
-			}
+		if (_routesCustom) {
```

Comment:
@jahands what happens if Advanced Mode projects don't provide a custom `_routes.json` file? How are worker "guarded" routes determined?

Reply:
If there is no `_routes.json`, don't send one. The backend will default to all routes running through Functions